### PR TITLE
[algo] feat: Implemented SAPO algorithm by Qwen

### PR DESCRIPTION
**Pull request by BounharAbdelaziz**
Implements **Soft Adaptive Policy Optimization (SAPO)** for RL fine-tuning of LLMs. SAPO replaces hard clipping with temperature-controlled soft gating for more stable training and better sample efficiency. Paper: https://arxiv.org/abs/2511.20347

### Checklist Before Starting
- [x] Search for similar PRs: [SAPO search](https://github.com/NovaSky-AI/SkyRL/pulls?q=is%3Apr+is%3Aopen+sapo)
- [x] Format PR title: `[algo] feat: Implemented SAPO algorithm by Qwen`

### Design & Code Changes
- Added `sapo_policy_loss()` in `skyrl-train/skyrl_train/utils/ppo_utils.py`
- Added `τ_neg` and `τ_pos` in `skyrl-train/skyrl_train/config/ppo_base_config.yaml`
- Implements soft gate: `f(r) = σ(τ(r-1)) · 4/τ` where `r = π_θ / π_θ_old`
- Uses asymmetric temperatures (`τ_neg > τ_pos`) as in original paper
- Aggregation: `sequence_mean` as per paper

### Test
- **Setup**: Qwen3-4B-Base, total batch size 256, mini-batch size 32, context len 8192.
- **PG loss**: The PG loss under SAPO stays around −0.01, i.e., the underlying SAPO objective (before the leading minus sign) is positive, indicating the model consistently discovers and reinforces high-reward behaviors.

image

### CODE REVIEW

**Comment by BounharAbdelaziz** [comment]
@tyler-griggs @erictang000

File: skyrl-train/examples/algorithms/sapo/run_dapo_aime_qwen3_4b_aime.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+N_SAMPLES_PER_PROMPT=16
+EVAL_N_SAMPLES_PER_PROMPT=32
+ENFORCE_EAGER=true # cuda graphs can cause some instability
+LR=1e-6
+
+uv run --isolated --extra vllm -m examples.algorithms.dapo.main_dapo \
+  data.train_data="['$TRAIN_FILE']" \
+  data.val_data="['$TEST_FILE']" \
+  trainer.algorithm.advantage_estimator="grpo" \
+  trainer.algorithm.policy_loss_type="dual_clip" \
```
Comment (erictang000) [inline skyrl-train/examples/algorithms/sapo/run_dapo_aime_qwen3_4b_aime.sh]:
```suggestion
  trainer.algorithm.policy_loss_type="sapo" \
```

Comment (erictang000) [reply skyrl-train/examples/algorithms/sapo/run_dapo_aime_qwen3_4b_aime.sh]: @BounharAbdelaziz, for the metrics in the PR description you pasted did you set this to sapo? Seeing that it's set correctly for the slurm script in this PR: https://github.com/NovaSky-AI/SkyRL/pull/766/files, just wondering

**Review by erictang000 — APPROVED** [review]
made some minor changes, and verified everything works, thanks @BounharAbdelaziz!

**Comment by BounharAbdelaziz** [comment]
> made some minor changes, and verified everything works, thanks @BounharAbdelaziz!

Thanks! Glad everything works on your side too :)
